src/java/org/apache/cassandra/service/opstate/KeyspaceTableOpStatePersister.java

```diff
@@ -0,0 +1,116 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.cassandra.service.opstate;
+
+import java.io.File;
+import java.io.IOException;
+import java.time.Instant;
+import java.util.HashMap;
+import java.util.Map;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import org.codehaus.jackson.map.ObjectMapper;
+
+public class KeyspaceTableOpStatePersister
+{
+    private static final Logger log = LoggerFactory.getLogger(KeyspaceTableOpStatePersister.class);
+
+    private static final ObjectMapper OBJECT_MAPPER = new ObjectMapper();
+
+    private File persistentFile;
+    private final String persistentFileLocation;
+
+    public KeyspaceTableOpStatePersister(String persistentFileLocation)
+    {
+        this.persistentFileLocation = persistentFileLocation;
+        this.persistentFile = getOrMaybeCreateStateFile();
+    }
+
+    public Map<KeyspaceTableKey, Instant> readStateFromFile()
+    {
+        File persistentStateFile = getOrMaybeCreateStateFile();
+        if (persistentStateFile == null || persistentStateFile.length() == 0)
+            return new HashMap<>();
+
+        try
+        {
+            return convertMapTypeToKeyspaceTableKeyInstant(OBJECT_MAPPER.readValue(persistentStateFile, Map.class));
+        }
+        catch (Exception e)
+        {
+            log.warn("Failed to retrieve state from file.", persistentStateFile.getAbsolutePath(), e);
+            return new HashMap<>();
```

Comment:
If we have a file on disk that we fail to read initially I think it's possible we could:
* populate the cache with an empty map
* the next cleanup update will have the cache keep a map with a single entry
* try to persist the cache to disk
* persister successfully reads the file this time
* overwrite the entire map with the single entry known by the cache

I think we'd have to reconcile the cache with what is on disk when we are eventually able to read the file successfully

Reply:
I can probably read before write and take the most updated of all entries. However, theoretically, I wonder if the above will be a truly problematic failure mode since the cache should always be more updated than the file. As a result, if my understanding is correct, you are only losing info (instead of having incorrect one). The worse thing that could happen in this case is the subsequent cleanup being delayed